src/Common/src/System/Marvin.cs

```diff
@@ -14,74 +14,53 @@ internal static class Marvin
         {
-            long hash64 = ComputeHash(ref data, count, seed);
+            long hash64 = ComputeHash(data, seed);
             return ((int)(hash64 >> 32)) ^ (int)hash64;
         }
 
         /// <summary>
         /// Computes a 64-hash using the Marvin algorithm.
         /// </summary>
-        public static long ComputeHash(ref byte data, int count, ulong seed)
+        public static long ComputeHash(ReadOnlySpan<byte> data, ulong seed)
         {
-            uint ucount = (uint)count;
             uint p0 = (uint)seed;
             uint p1 = (uint)(seed >> 32);
 
-            int byteOffset = 0;  // declared as signed int so we don't have to cast everywhere (it's passed to Unsafe.Add() and used for nothing else.)
-
-            while (ucount >= 8)
+            if (data.Length >= sizeof(uint))
             {
-                p0 += Unsafe.As<byte, uint>(ref Unsafe.Add(ref data, byteOffset));
-                Block(ref p0, ref p1);
+                ReadOnlySpan<uint> uData = data.NonPortableCast<byte, uint>();
 
-                p0 += Unsafe.As<byte, uint>(ref Unsafe.Add(ref data, byteOffset + 4));
-                Block(ref p0, ref p1);
+                for (int i = 0; i < uData.Length; i++)
+                {
+                    p0 += uData[i];
+                    Block(ref p0, ref p1);
+                }
 
-                byteOffset += 8;
-                ucount -= 8;
+                // byteOffset = data.Length - data.Length % 4
+                // is equivalent to clearing last 2 bits of length
+                // Using it directly gives a perf hit for short strings making it at least 5% or more slower.
+                int byteOffset = data.Length & (~3);
+                data = data.Slice(byteOffset);
             }
 
-            switch (ucount)
+            switch (data.Length)
             {
-                case 4:
-                    p0 += Unsafe.As<byte, uint>(ref Unsafe.Add(ref data, byteOffset));
-                    Block(ref p0, ref p1);
-                    goto case 0;
-
                 case 0:
                     p0 += 0x80u;
                     break;
 
-                case 5:
-                    p0 += Unsafe.As<byte, uint>(ref Unsafe.Add(ref data, byteOffset));
-                    byteOffset += 4;
-                    Block(ref p0, ref p1);
-                    goto case 1;
-
                 case 1:
-                    p0 += 0x8000u | Unsafe.Add(ref data, byteOffset);
+                    p0 += 0x8000u | data[0];
                     break;
 
-                case 6:
-                    p0 += Unsafe.As<byte, uint>(ref Unsafe.Add(ref data, byteOffset));
-                    byteOffset += 4;
-                    Block(ref p0, ref p1);
-                    goto case 2;
-
                 case 2:
-                    p0 += 0x800000u | Unsafe.As<byte, ushort>(ref Unsafe.Add(ref data, byteOffset));
+                    p0 += 0x800000u | data.NonPortableCast<byte, ushort>()[0];
```

Comment:
`NonPortableCast` does  relatively heavyweight calculations.
It may be faster to do
```C#
p0 += 0x800000u | data[0] << 8 | data[1];
```

Reply:
Thanks @omariom. We have just opened 2 JIT issues related to this:
https://github.com/dotnet/coreclr/issues/15492 (non portable cast)
https://github.com/dotnet/coreclr/issues/15491 (other similar)

hopefully we can fix it at the source :-)

Reply:
@krwq NonPortable cast is slow and heavy weight by design. There is not much JIT can do about it. The code inside this switch statement should be change to what @omariom suggested.

Reply:
Oops, completely forgot about that issue. Yes, it's problematic for the JIT to optimize `NonPortableCast` but we can optimize the C# code manually.

Reply:
And it would be better to use `uint BitConverter.ToUInt32(ReadOnlySpan<byte> value)` to read the uints instead of NonPortableCast to make this code fully portable (though that may require JIT fixes to get equivalent performance).

Reply:
@jkotas, at tip of master branch this has been changed to `p0 += 0x800000u | MemoryMarshal.Cast<byte, ushort>(data)[0];`, is NonPortable perf penalty issue mitigated with `MemoryMarshal.Case<T1,T2>`?

Reply:
No. NonPortable -> MemoryMarshal.Cast was a simple rename.

Reply:
Thanks, I have sent a PR https://github.com/dotnet/corefx/pull/28110.